### PR TITLE
Add authorize Monitor connection elements. 

### PR DESCRIPTION
**Pull request by anastasia-malysheva**
…Server chain element. Add options to enable mointor authorize element. Add unit tests. Turn off authorize monitor by default in the network service chain.

Signed-off-by: anastasia.malysheva <anastasia.malysheva@xored.com>

## Issue link
https://github.com/networkservicemesh/sdk/issues/46

## How Has This Been Tested?

- [x] Added unit testing to cover
- [x] Tested manually
- [ ] Tested by integration testing
- [ ] Have not tested

## Types of changes

- [ ] Bug fix
- [x] New functionallity
- [ ] Documentation
- [ ] Refactoring
- [ ] CI

### CODE REVIEW

File: pkg/networkservice/chains/nsmgrproxy/server.go [hunk 1]
```diff
@@ -117,6 +119,16 @@ func WithAuthorizeServer(authorizeServer networkservice.NetworkServiceServer) Op
 	}
 }
 
+// WithAuthorizeMonitorServer sets authorization MonitorConnectionServer chain element
+func WithAuthorizeMonitorServer(authorizeMonitorServer networkservice.MonitorConnectionServer) Option {
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgrproxy/server.go:123]: Please apply for other places

```suggestion
func WithAuthorizeMonitorConnectionServer(authorizeMonitorServer networkservice.MonitorConnectionServer) Option {
```

File: pkg/tools/monitorconnection/authorize/server.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+				},
+			)
+			return true
+		},
+	)
+
+	seg := make([]string, 0)
+	for _, v := range in.PathSegments {
+		seg = append(seg, v.GetId())
+	}
```
Comment (denis-tingaikin) [inline pkg/tools/monitorconnection/authorize/server.go:77]: Please check  only previous segment.

File: pkg/tools/opa/policies/service_connection.rego [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+package nsm
+
+default service_connection = false
+
+service_connection {
+   conn_ids := [y | y = input.spiffe_id_connection_map[input.service_spiffe_id][_]]
+   count(input.path_segments) > 0
+   count(conn_ids) > 0
+   conn_ids == input.path_segments
+}
```
Comment (denis-tingaikin) [inline pkg/tools/opa/policies/service_connection.rego:26]: Please note: the client is always previous segment. Please check only previous id.

File: pkg/tools/opa/opainput.go [hunk 4]
```diff
@@ -55,7 +54,8 @@ func pemEncodingX509Cert(cert *x509.Certificate) string {
 	return string(certpem)
 }
 
-func parseX509Cert(authInfo credentials.AuthInfo) *x509.Certificate {
+// ParseX509Cert - parses x509 certificate from the passed credentials.AuthInfo
+func ParseX509Cert(authInfo credentials.AuthInfo) *x509.Certificate {
```
Comment (denis-tingaikin) [inline pkg/tools/opa/opainput.go:58]: Can be private

File: pkg/networkservice/common/authorize/server.go [hunk 5]
```diff
@@ -29,40 +29,59 @@ import (
...
 	}
+
+	if spiffeID, err := spire.SpiffeIDFromContext(ctx); err == nil {
+		connID := conn.GetPath().GetPathSegments()[index-1].GetId()
+		ids, ok := a.spiffeIDConnectionMap.Load(spiffeID)
+		if !ok {
+			ids = &spire.ConnectionIDSet{}
+		}
+		var placer struct{}
+		ids.Store(connID, placer)
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/authorize/server.go:82]:
```suggestion
		ids.Store(connID, struct{}{})
```

Comment (anastasia-malysheva) [reply pkg/networkservice/common/authorize/server.go]: fixes

File: pkg/tools/opa/policies/service_connection.rego (hunk 3 above)
Comment (anastasia-malysheva) [reply pkg/tools/opa/policies/service_connection.rego]: fixed

File: pkg/tools/opa/opainput.go (hunk 4 above)
Comment (anastasia-malysheva) [reply pkg/tools/opa/opainput.go:58]: I use this method in other package.

File: pkg/tools/monitorconnection/next/tail_client.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2022 Cisco and/or its affiliates.
```
Comment (denis-tingaikin) [inline pkg/tools/monitorconnection/next/tail_client.go:1]: this file  can be removed

Comment (anastasia-malysheva) [reply pkg/tools/monitorconnection/next/tail_client.go]: removed

**Review by denis-tingaikin — APPROVED** [review]

**Pull request merged**